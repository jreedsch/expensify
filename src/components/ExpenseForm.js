import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css';

//const now = moment();  today
//console.log(now.format('MMM Do, YYYY'));

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100) : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      err: ''
    };

  };

  onDescriptionChange = (e) => {
    const description = e.target.value; // or use event.persist()
    this.setState(() => ({ description }));  //es6 object shorthand
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));  //es6 object shorthand
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    const currencyRegex = '/^\d{1,}(\.\d{0,2})$/'; // ^ begins with
                                              // \d decimal
                                              // * any count, {1,} 1 to any
                                              // () optional
                                              // decimal point (escaped)
                                              // {0,2} from 0 to 2 decimals
                                              // $ end
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));  //es6 object shorthand
    } else {
      console.log("amount does not match regex filter: "+amount);
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onCalendarFocusChange = ({focused}) => {
    this.setState(() => ({ calendarFocused: focused}));
  };



  onSubmitHandler = (e) => {
    console.log("SUBMIT, description="+this.state.description);
    e.preventDefault(); //no refresh
    if (!this.state.description || !this.state.description.length > 0) {
    //if (!this.state.description || !this.state.amount > 0) {
      this.setState(() => ({ err: 'description is required' }));
    } else if (!this.state.amount || !this.state.amount > 0) {
    //if (!this.state.description || !this.state.amount > 0) {
      this.setState(() => ({ err: 'amount is required' }));
    } else {
      this.setState(() => ({ err: '' }));
      console.log("IN ExpenseForm, state: "+JSON.stringify(this.state));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }

  }

  render() {
    return (

        <form className="form" onSubmit={this.onSubmitHandler}>
          {this.state.err && <p className="form__error">{this.state.err}</p> }
          <input
            type='text'
            className="text-input"
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type='text'
            placeholder='Amount'
            className="text-input"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            className="textarea-input"
            placeholder='Note'
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <div> { /*  removes from form children (_form.scss: > * { lecture #176 ), so not full width  */ }
            <span><button type="submit" >Save Expense</button></span>
            <span><button type="button" className='cancel_button' onClick={this.props.onCancel}>Cancel</button></span>
          </div>
        </form>
    )

  }
}

export default ExpenseForm ;
