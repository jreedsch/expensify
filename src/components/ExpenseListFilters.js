import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount,sortByDate, setStartDate, setEndDate } from  '../actions/filters';
import { DateRangePicker } from 'react-dates';

// dispatch is set on props by react-redux
export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    console.log("IN ExpenseListFilters.onFocusChange");
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    //const ndx = e.target.selectedIndex; //can use e.target.value
    //if (ndx === 0) {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>

        filter by description:
        <input
          type='text'
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />

        sort by:
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
  //???  >
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>

        show date range:
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    ); // end return
  }; //end render
}; //end class

const mapDispatchToProps = (dispatch, props) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

const mapStateToProps = (state) => ( {filters: state.filters} ); //implicit return
//const mapStateToProps = (state) => {   //explicit return
//  return {
//    filters: state.filters
//  };
//};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
