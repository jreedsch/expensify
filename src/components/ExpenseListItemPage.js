import React from 'react';
//import { connect } from 'react-redux';
//import { removeExpense } from '../actions/expenses';
import { Link }  from 'react-router-dom';
import moment from 'moment';

//const ExpenseListItemPage = (props) => ( //implicit return
//const ExpenseListItemPage = ({dispatch, description, amount, createdAt, id}) => { //destructured, from caller's spread operator
const ExpenseListItemPage = ({description, amount, createdAt, id}) => { //destructured, from caller's spread operator
//export const ExpenseListItemPage = ({description, amount, createdAt, id}) => { //destructured, from caller's spread operator
  const createdAtMoment = moment(createdAt).format("MMM Do YYYY");
  return (
    <div>
      <p><Link to={`/edit/${id}`}>{description}</Link></p>
      <p>amount: {amount}, createdAt: {createdAtMoment}</p>



    </div>
  );
};

//      <button onClick={(event) => {dispatch(removeExpense({id}))}}>Remove!</button>

//    <p>description: {props.expense.description} amount: {props.expense.amount} createdAt: {props.expense.createdAt}</p>


// just get dispatch from connect
//export default connect()(ExpenseListItemPage);
export default ExpenseListItemPage;
