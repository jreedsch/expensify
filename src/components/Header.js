import React from 'react';
import { NavLink }  from 'react-router-dom';

const Header = () => ( //implicit return
  <header>
    <h1>Expensify</h1>
    <div><NavLink to='/' activeClassName='is-active' exact={true}>Home</NavLink></div>
    <div><NavLink to='/create' activeClassName='is-active'>Add</NavLink></div>
    <div><NavLink to='/help'  activeClassName='is-active'>Help</NavLink></div>
  </header>
);

export default Header;
