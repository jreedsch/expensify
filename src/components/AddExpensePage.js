import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses'; //addExpense

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    console.log("IN AddExpensePage, add submitted.");
    this.props.startAddExpense(expense); //props.dispatch(addExpense(expense)); see mapDispatchToProps
    this.props.history.push('/');
  };

  render() {
    return (
        <div>
          <h3>add a new expense page</h3>
          <ExpenseForm onSubmit={this.onSubmit}/>
        </div>
    );
  }
}

/*
const AddExpensePage = (props) => ( //implicit return
  <div>
    <h3>add expense page</h3>
    <ExpenseForm onSubmit={(expense) => {
      console.log("IN AddExpensePage, add submitted.");
      props.onSubmit(expense); //props.dispatch(addExpense(expense)); see mapDispatchToProps
      props.history.push('/');
    }}/>
  </div>
);
*/

// makes testing the dispatch much easier
const mapDispatchToProps = (dispatch) => ({startAddExpense: (expense) => dispatch(startAddExpense(expense)) })
//{
//  return {
//    onSubmit: (expense) => dispatch(addExpense(expense))
//  }
//};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
