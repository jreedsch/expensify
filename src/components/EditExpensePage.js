import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

  onSubmitEdit = (expense) => {
    console.log("IN EditExpensePage, edit submitted, expense: "+JSON.stringify(expense));
    console.log("this.props 1: "+JSON.stringify(this.props));
    console.log("edit expense to submit:  "+JSON.stringify(expense));
    console.log("edit expense id to submit:  "+this.props.expense.id);
    this.props.editExpense(this.props.expense.id, expense); //(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemoveButtonClick = (event) => {
    console.log("this.props 2: "+JSON.stringify(this.props));
    this.props.removeExpense({id: this.props.expense.id}); //this.props.removeExpense(this.props.expense.id)
    this.props.history.push('/');
  };

//        <p> edit expense page for id {this.props.match.params.id}</p>
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmitEdit}
        />
        <button onClick={ this.onRemoveButtonClick } >Remove!</button>
      </div>
    )
  };
}

const mapDispatchToProps = (dispatch, props) => ({
  removeExpense: (data) => dispatch(startRemoveExpense(data)), //dispatch(removeExpense(data)),
  editExpense: (id, expense) => dispatch(startEditExpense(id, expense))
});
  //removeExpense: (id) => dispatch(removeExpense({ id })) }

//const mapDispatchToProps = (dispatch) => ([
//  {editExpense: (id, expense) => dispatch(editExpense(id, expense)) },
//  {removeExpense: (id) => dispatch(removeExpense({ id })) }
//]);

/*
const EditExpensePage = (props) => {
  console.log("IN EditExpensePage, props: "+JSON.stringify(props)); //props from react-router
  return (
    <div>
      <p> edit expense page for id {props.match.params.id}</p>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          console.log("IN EditExpensePage, edit submitted, expense: "+JSON.stringify(expense));
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
      <button onClick={(event) => {
        props.dispatch(removeExpense({id: props.expense.id}));
        props.history.push('/');
      }}>Remove!</button>
    </div>
  );
};

*/

//$$$ const mapStateToProps = (state, props) => ( { expense: state.expenses.find((expense) =>  expense.id === props.match.params.id) } );
const mapStateToProps = (state, props) => {
  const editExpense = state.expenses.find((expense) =>  expense.id === props.match.params.id);
  console.log("IN EditExpensePage.mapStateToProps, state: "+JSON.stringify(state));
  console.log("IN EditExpensePage.mapStateToProps, props: "+JSON.stringify(props));
  console.log("IN EditExpensePage.mapStateToProps, editExpense: "+JSON.stringify(editExpense));
  return { expense:  editExpense };
};



/*
const mapStateToProps = (state, props) => {
  return {
    //expense: state.expenses.find((expense) => {
    //  return expense.id === props.match.params.id;
    //})
    expense: state.expenses.find((expense) =>  expense.id === props.match.params.id)
  };
};
*/

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage) ;
