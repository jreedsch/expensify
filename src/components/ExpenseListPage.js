import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItemPage from './ExpenseListItemPage';
import { getVisibleExpenses, getExpenseTotal } from '../selectors/expenses';

// need 2 exports: unconnected version (here) from Jest, connected version(default) for runtime
export const ExpenseListPage = (props) => ( //implicit return
  <div className='content-container'>
    { /* <h3>expense list page for filter: {props.filters.text}</h3> */ }
    { /* }<h3>count: {props.expenses.length}   amount: {getExpenseTotal(props.expenses)} </h3> */ }
    <div className='list-header'>
       <div className='show-for-mobile'> {props.expenses.length} Expenses</div> { /* mobile only */}
       <div className='show-for-desktop'>Expense (count: {props.expenses.length})</div>
       <div className='show-for-desktop'>Amount</div>
    </div>
    <div className='list-body'>
      {props.expenses.length === 0
        ? (<div className='list-item__empty'>
            <span>No expenses</span>
           </div>)
        : ( props.expenses.map((expense) => {
              return <ExpenseListItemPage key={expense.id} {...expense}  />;
            })
          )
      }
    </div>
  </div>
);
//       ( <ExpenseListItemPage key={ndx} expense={expense} ndx={ndx} /> )

// this is a HOC

/*
const ConnectedExpenseListPage = connect((state) => {
  return {expenses: state.expenses};
})(ExpenseListPage);

export default ConnectedExpenseListPage;
*/

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters),
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListPage);
