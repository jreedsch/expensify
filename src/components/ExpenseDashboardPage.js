import React from 'react';
import ExpenseListPage from './ExpenseListPage';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashboardPage = () => ( //implicit return
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseListPage />
    <p style={styles} >src/components/ExpenseDashboardPage.js</p>
  </div>
);

const styles = {
  fontSize: '10px',
  float: 'right'
}
export default ExpenseDashboardPage ;
