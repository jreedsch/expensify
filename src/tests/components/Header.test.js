//import ReactShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import React from 'react';
//import toJSON from 'enzyme-to-json'; // see jest.config.json
import { Header } from '../../components/Header';
//import { startLogout } from '../../actions/auth';

// in Jest, 'u' to save new baseline snapshot
test('should render Header correctly', () => {
  const startLogoutMock = jest.fn();
  const wrapper = shallow(<Header startLogout={startLogoutMock}/>);
  //expect(toJSON(wrapper)).toMatchSnapshot();
  expect(wrapper).toMatchSnapshot();
  //expect(wrapper.find('h1').length).toBe(1); //count of h1 elements
  //expect(wrapper.find('h1').text()).toBe('Expensify'); //contents of h1 element

  // using ReactShallowRenderer
  //const renderer = new ReactShallowRenderer();
  //renderer.render(<Header />);
  //console.log(renderer.getRenderOutput());
  //expect(renderer.getRenderOutput()).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
  const startLogoutMock = jest.fn();  //spy
  const wrapper = shallow(<Header startLogout={startLogoutMock} />);
  wrapper.find('button').simulate('click');
  expect(startLogoutMock).toHaveBeenCalled();
});
