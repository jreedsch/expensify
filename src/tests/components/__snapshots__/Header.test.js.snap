// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <div>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      exact={true}
      to="/dashboard"
    >
      Dashboard
    </NavLink>
  </div>
  <div>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      to="/create"
    >
      Add
    </NavLink>
  </div>
  <div>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      to="/help"
    >
      Help
    </NavLink>
  </div>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</header>
`;
